app/services/IntegrationFrameworkService.scala

```diff
@@ -16,78 +16,94 @@
+    val response = for {
+      benefitId <- createOrUpdateStateBenefit(stateBenefitsUserData)
+      result <- createOrUpdateBenefitDetails(stateBenefitsUserData, benefitId)
+    } yield result
+    response.value
   }
 
-  def removeOrIgnoreClaim(userData: StateBenefitsUserData, benefitId: UUID)
+  def removeOrIgnoreClaim(stateBenefitsUserData: StateBenefitsUserData)
                          (implicit hc: HeaderCarrier): Future[Either[ApiServiceError, Unit]] = {
-    val eventualIgnoreOrDelete = if (userData.isHmrcData) {
-      integrationFrameworkConnector.ignoreStateBenefit(userData.taxYear, userData.nino, benefitId)
+    val benefitId = stateBenefitsUserData.claim.get.benefitId.get
+    val eventualIgnoreOrDelete = if (stateBenefitsUserData.isHmrcData) {
+      connector.ignoreStateBenefit(stateBenefitsUserData.taxYear, stateBenefitsUserData.nino, benefitId)
     } else {
-      integrationFrameworkConnector.deleteStateBenefit(userData.taxYear, userData.nino, benefitId)
+      connector.deleteStateBenefit(stateBenefitsUserData.taxYear, stateBenefitsUserData.nino, benefitId)
     }
+
     eventualIgnoreOrDelete.map {
       case Left(error) => Left(ApiServiceError(error.status.toString))
       case Right(_) => Right(())
     }
   }
 
-  def unIgnoreClaim(userData: StateBenefitsUserData)
+  def unIgnoreClaim(stateBenefitsUserData: StateBenefitsUserData)
                    (implicit hc: HeaderCarrier): Future[Either[ApiServiceError, Unit]] = {
-    val benefitId = userData.claim.get.benefitId.get
-    integrationFrameworkConnector.unIgnoreStateBenefit(userData.taxYear, userData.nino, benefitId).map {
+    val benefitId = stateBenefitsUserData.claim.get.benefitId.get
+    connector.unIgnoreStateBenefit(stateBenefitsUserData.taxYear, stateBenefitsUserData.nino, benefitId).map {
       case Left(error) => Left(ApiServiceError(error.status.toString))
       case Right(_) => Right(())
     }
   }
 
-  private def createOrUpdateStateBenefit(userData: StateBenefitsUserData, claimData: ClaimCYAModel)
-                                        (implicit hc: HeaderCarrier): Future[Either[ApiError, UUID]] = if (userData.isNewClaim) {
-    integrationFrameworkConnector.addStateBenefit(userData.taxYear, userData.nino, AddStateBenefit(userData.benefitType, claimData))
-  } else {
-    val benefitId = claimData.benefitId.get
-    integrationFrameworkConnector.updateStateBenefit(userData.taxYear, userData.nino, benefitId, UpdateStateBenefit(claimData)).map {
-      case Left(error) => Left(error)
+  private def createOrUpdateStateBenefit(userData: StateBenefitsUserData)
+                                        (implicit hc: HeaderCarrier): EitherT[Future, ApiServiceError, UUID] = userData match {
+    case _ if userData.isHmrcData => EitherT(Future.successful[Either[ApiServiceError, UUID]](Right(userData.claim.get.benefitId.get)))
+    case _ if userData.isNewClaim => addCustomerStateBenefit(userData)
+    case _ => updateCustomerStateBenefit(userData)
```

Comment:
Not sure how valuable the use of a match with guards is over regular if elses. I guess we use fewer lines...

Reply:
Yes, and it is more readable